### PR TITLE
server: fix edge case in tracing tenant requests

### PR DESCRIPTION
**Pull request by andreimatei**
The tracing of tenant Batch RPCs has special handling for trace
redaction when the request comes from a tenant. There was a problem with
it though, and the code was also quite confusing. Dealing with tenant
requests was done a little too brutal; the code was equating such
requests with non-local RPCs. Direct RPCs from tenants are non-local by
definition, but they can spawn child RPCs, and these child RPCs can be
local. In such a case, the code was erroneously handling the child RPC -
we were not creating a span for it as we should have, and then we were
redacting the parent span, even though the redaction of the parent will
happen again later on return from the parent RPC.

Release note: None

### CODE REVIEW

File: pkg/server/node.go [hunk 1]
```diff
@@ -1032,20 +1034,30 @@ func (n *Node) setupSpanForIncomingRPC(
...
+				// needs to be redacted if the response goes to a tenant. In case the request
+				// is local, then the trace might eventually go to a tenant (and tenID might
+				// be set), but it will go to the tenant only indirectly, through the response
+				// of a parent RPC. In that case, that parent RPC is responsible for the
+				// redaction.
+				//
+				// Tenants get a redacted recording, i.e. with anything
+				// sensitive stripped out of the verbose messages. However,
+				// structured payloads stay untouched.
+				needRedaction := tenID == roachpb.SystemTenantID && !localRequest
```
Comment (tbg) [inline pkg/server/node.go:1053]: `!=`

Looks like `TestTenantTracesAreRedacted/regular-tenant` caught this, too.

Comment (andreimatei) [reply pkg/server/node.go]: done, thanks

**Review by dhartunian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73208)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @dhartunian, and @tbg)

___
*[pkg/server/node.go, line 1016 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73208#-Mpi1VHRCCQHmtvBh06r:-Mpi1VHRCCQHmtvBh06s:b-nj555l) ([raw file](https://github.com/cockroachdb/cockroach/blob/37f5ffbb110d94709b219c780b84a57dff2d064c/pkg/server/node.go#L1016)):*
> ```Go
> 	parentSpan := tracing.SpanFromContext(ctx)
> 	localRequest := grpcutil.IsLocalRequestContext(ctx)
> 	if localRequest {
> ```

Just trying to understand this change, why can we remove the `tenID == roachpb.SystemTenantID` check here? Are all local requests using the system tenant by definition?
___
*[pkg/server/node.go, line 1053 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73208#-Mpi0hWg2nmECbPbk3L9:-Mpi0hWg2nmECbPbk3LA:b-qwjne5) ([raw file](https://github.com/cockroachdb/cockroach/blob/37f5ffbb110d94709b219c780b84a57dff2d064c/pkg/server/node.go#L1053)):*
> ```Go
> 				// sensitive stripped out of the verbose messages. However,
> 				// structured payloads stay untouched.
> 				needRedaction := tenID != roachpb.SystemTenantID && !localRequest
> ```

is `&& !localRequest` redundant here? You already have it above in the `if` statement.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73208)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @tbg)

___
*[pkg/server/node.go, line 1016 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73208#-Mpi1VHRCCQHmtvBh06r:-MpiCeHT4lWR7a4cb-bF:b-7lyus7) ([raw file](https://github.com/cockroachdb/cockroach/blob/37f5ffbb110d94709b219c780b84a57dff2d064c/pkg/server/node.go#L1016)):*
> Are all local requests using the system tenant by definition?

No, but they probably should be: https://cockroachlabs.slack.com/archives/C02J7BTSRK2/p1637972742007000

> why can we remove the tenID == roachpb.SystemTenantID check here

Well, the check here was just not making sense. When `tenID != roachpb.SystemTenantID`, we were behaving as if this was not a local request, which was bad because we were not opening a span for it. We want that span; creating it or not has nothing to do with the request being for a tenant or not.

With the check gone, local requests with a tenant id no longer get any redaction. You may be asking  why that's OK, and the answer is that the redaction will come at the process boundary - when the parent RPC returns to the (tenant) client, that's when we'll redact the whole trace, including this span and it's children. Before this patch, in cases of nested RPCs, we were redacting parts twice.

Makes sense?
___
*[pkg/server/node.go, line 1053 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73208#-Mpi0hWg2nmECbPbk3L9:-MpiCQNU4ZhewOOEgra6:b-ei59qz) ([raw file](https://github.com/cockroachdb/cockroach/blob/37f5ffbb110d94709b219c780b84a57dff2d064c/pkg/server/node.go#L1053)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

is `&& !localRequest` redundant here? You already have it above in the `if` statement.
</blockquote></details>

yup, done

**Review by dhartunian — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73208#-:-MplE5_2DcezDbm8JhW9:bks4kj3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @dhartunian, and @tbg)

___
*[pkg/server/node.go, line 1016 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/73208#-Mpi1VHRCCQHmtvBh06r:-MplDgEEAK5nML38xFFk:bn52xl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/37f5ffbb110d94709b219c780b84a57dff2d064c/pkg/server/node.go#L1016)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> Are all local requests using the system tenant by definition?

No, but they probably should be: https://cockroachlabs.slack.com/archives/C02J7BTSRK2/p1637972742007000

> why can we remove the tenID == roachpb.SystemTenantID check here

Well, the check here was just not making sense. When `tenID != roachpb.SystemTenantID`, we were behaving as if this was not a local request, which was bad because we were not opening a span for it. We want that span; creating it or not has nothing to do with the request being for a tenant or not.

With the check gone, local requests with a tenant id no longer get any redaction. You may be asking  why that's OK, and the answer is that the redaction will come at the process boundary - when the parent RPC returns to the (tenant) client, that's when we'll redact the whole trace, including this span and it's children. Before this patch, in cases of nested RPCs, we were redacting parts twice.

Makes sense?
</blockquote></details>

:thumbsup: thx for the explanation.

**Pull request merged**